### ISSUE TITLE
Reflection: train docs/sdlc/*.md from session transcripts (backpass-inspired evidence loop)

### ISSUE THREAD

**Issue by valorengels** [tomcounsell/ai#3062]
> **Context for the reader.** This repo runs a scheduled *reflections* system (`reflections/`, configured in `~/Desktop/Valor/reflections.yaml`, 33 entries) — background jobs that inspect the system's own history and act on what they find. It also runs an *SDLC pipeline* of skills (`/do-plan`, `/do-build`, `/do-pr-review`, `/do-merge`, …) whose behavior is defined by two layers of instruction files: generic bodies in `.claude/skills-global/do-*/SKILL.md` (hardlinked to every machine) and repo-specific addenda in [`docs/sdlc/*.md`](../tree/main/docs/sdlc). This issue proposes a new reflection that closes the loop between the two: the pipeline generates transcripts, but nothing today reads those transcripts back into the instruction files.
>
> The inspiration is [kunchenguid/backpass](https://github.com/kunchenguid/backpass), an open-source tool that treats agent memory files as trainable weights. Its tagline: *"You don't write AGENTS.md. You train it with gradient descent."*

## Problem

**Our SDLC instruction surface has no feedback path.** `docs/sdlc/do-build.md`, `docs/sdlc/do-plan.md`, and their siblings are written by hand, edited when someone remembers, and never validated against what actually happened in the sessions they governed. When a pipeline stage repeatedly goes wrong the same way — a builder that keeps skipping a gate, a plan stage that keeps under-specifying test scope — that lesson lands in a human's head or a memory note, not in the file the next agent will read.

**Current behavior.** Two existing reflections come close and both stop short:

1. `reflections/session_intelligence.py` is the only transcript-mining loop we have. It reads yesterday's `AgentSession` logs, runs an LLM classifier that emits one of six categories — `misunderstanding`, `code_bug`, `poor_planning`, `tool_misuse`, `scope_creep`, `integration_failure` (`reflections/utilities.py:454-455`) — and then **discards five of the six**. The gate is explicit:

   ```python
   # reflections/utilities.py:568
   if reflection.get("category") != "code_bug":
       return False
   ```

   The docstring says this is deliberate (#1414): non-`code_bug` categories "describe agent behaviour, not code defects, and must not ride into the auto-filer on length alone." That reasoning is sound *for the output channel it guards* — the auto-filer only knows how to open a `bug` issue. But it means `poor_planning`, `tool_misuse`, and `scope_creep` findings — precisely the signal about how our SDLC skills misbehave — are computed every night and thrown away.

2. `reflections/audits/skills_audit.py` (`reflections.auditing.run_skills_audit`) checks SKILL.md files for *structural* defects and files an issue when a finding survives 2 consecutive runs. It reads the instruction files but never reads a transcript, so it can tell you a section is missing but never that a present section is wrong.

**Nothing reads session transcripts and proposes edits to the instruction files.** Confirmed: no doc, plan, branch, issue, or commit in this repo mentions backpass or an equivalent mechanism.

**Desired outcome.** A scheduled reflection that treats recurring SDLC failures as evidence, accumulates that evidence across runs, and proposes concrete, human-gated edits to `docs/sdlc/*.md` via a PR — so the instruction surface improves from observed behavior instead of from memory.

## Definitions

| Term | Meaning |
|---|---|
| **backpass** | [kunchenguid/backpass](https://github.com/kunchenguid/backpass) — the inspiration. A five-stage local-first pipeline: collect transcripts → distill (96-99% reduction) → per-transcript "loss" call (which instructions helped, which were violated, what gaps exist, every claim carrying a verbatim quote) → aggregate gradients across sessions → synthesis model proposes edits in a staging dir, gated by evidence thresholds and a human `apply` step. |
| **Instruction surface** | The set of files an SDLC agent reads to know how to behave: `.claude/skills-global/do-*/SKILL.md` (global) and `docs/sdlc/*.md` (repo-specific addenda). |
| **Reflection** | A scheduled background job in `reflections/`, registered by dotted callable in `~/Desktop/Valor/reflections.yaml`. |
| **Gap ledger** | backpass's persistent record of gaps sighted across runs, so a lesson seen once is held until it recurs rather than acted on immediately. |
| **`docs_auditor` shape** | [`reflections/docs_auditor.py`](../blob/main/reflections/docs_auditor.py) — this repo's precedent for a reflection that *writes repo files*: deterministic detectors → dedicated `docs-audit/*` branch → PR → Redis dedup + daily PR cap + run lock → a paired sweeper reflection (`run_docs_branch_sweeper`) that closes stale branches/PRs after 14 days. |

## Solution sketch

A new reflection, `reflections/sdlc_backpass.py`, shaped like `docs_auditor` and fed by the classifier `session_intelligence` already runs.

**Pipeline (five stages, mirroring backpass):**

1. **Collect** — session transcripts from the last N days, bounded by our 30-day retention (`tools/disk_reclaim.py::DEFAULT_TRANSCRIPT_MAX_AGE_DAYS = 30`). Scope to SDLC-bearing sessions (those with a lane slug / `session/{slug}` branch), not conversational chatter.
2. **Distill** — collapse tool calls to one line each, truncate outputs, redact secrets. Non-negotiable: raw transcripts are almost entirely tool noise and would blow both cost and context.
3. **Score** — one cheap-model call per distilled transcript returning strict JSON: which `docs/sdlc/*.md` instructions were followed, which were violated, and what mistakes no instruction covers. **Every claim must carry a verbatim quote from the transcript.** Cache per transcript keyed on content hash + instruction-surface hash.
4. **Aggregate** — group evidence by instruction and by gap. A gap needs sightings in **≥2 distinct sessions** before it is eligible; proposing *removal* of an existing instruction needs harm-class evidence (following it caused damage) from ≥2 sessions — non-compliance alone never justifies deletion. Persist the ledger so sightings accumulate across nightly runs.
5. **Propose** — a high-reasoning model drafts edits in a staging copy, then the reflection opens a PR against a dedicated `sdlc-backpass/*` branch with the diff, the evidence quotes, and their session sources in the body. **The PR is the human gate. The reflection never writes to `main`.**

### Why this shape rather than widening the `session_intelligence` gate

The cheaper-looking change is to relax `is_high_confidence` so `poor_planning` / `tool_misuse` / `scope_creep` also clear the gate. **That is the wrong fix and should be explicitly rejected in the plan**, for three reasons:

- **Wrong output channel.** Widening the gate routes behavioral findings into an auto-filer that only knows how to open a `bug`-labeled GitHub issue. That produces a nightly stream of "the agent planned poorly" issues nobody can action — which is the failure mode #1414 already closed. The value of a backpass loop is not *noticing* the mistake; it is *proposing the specific instruction edit that prevents it*, with a diff a human can accept or reject.
- **Wrong evidence bar.** The current gate fires on a single session. Instruction edits must not. The ≥2-distinct-session threshold, the harm-class requirement for deletions, and a persistent ledger are the whole point, and none of them fit inside a boolean predicate.
- **No memory of rejection.** A per-run filer re-proposes whatever it re-sees. An instruction-editing loop must remember rejected edits and not resurface them without materially new evidence.

**However**, the two should be wired together rather than duplicated: the classifier in `reflections/utilities.py::run_llm_reflection` already computes exactly the categories this reflection needs. The plan should extract that output as a shared evidence stream so the discarded categories feed the new reflection, and leave `is_high_confidence` **unchanged** — `session_intelligence` remains the code-bug filer.

## Constraints (hard — the plan must honor these)

- **Never edit `.claude/skills-global/do-*/SKILL.md`.** `scripts/update/hardlinks.py::sync_claude_dirs` hardlinks that directory to `~/.claude/skills/` on every machine, so a global SKILL.md edit changes agent behavior on every machine, for every project, immediately and without review. This reflection targets **`docs/sdlc/*.md` only** — the repo-specific addendum layer. Valor-specific lessons have no business in a generic skill body. (See CLAUDE.md → "Global vs. Project-Only Skills", and [`docs/features/skill-context-convention.md`](../blob/main/docs/features/skill-context-convention.md).)
- **Do not break the skills audit.** `reflections.auditing.run_skills_audit` enforces required SKILL.md sections. Any proposed edit must leave those structural invariants intact; a backpass PR that trips the skills audit is a defect in this reflection, not in the audit.
- **Token budget.** `docs/sdlc/*.md` is loaded by SDLC agents at runtime. Adopt backpass's budget discipline: a per-file token budget, and every proposed ADD is measured against it. An instruction surface that only grows is a regression. Growth beyond budget must force a REMOVE or a REWRITE, not an override.
- **Dry-run first.** Follow `docs_auditor`: default to proposing-without-PR until a run has been observed producing sane output. Ship with the write path off.
- **Lock, dedup, and a cap.** Reuse the `docs_auditor` Redis idiom — a run lock, a per-finding dedup key, and a hard cap of at most one PR per day. Pair it with branch/PR sweeping so abandoned `sdlc-backpass/*` branches do not accumulate.
- **Never co-authored.** Commits and PRs from this reflection carry no `Co-Authored-By` or generated-by footer.

## Acceptance criteria

- [ ] A new reflection module exists and is registered in `~/Desktop/Valor/reflections.yaml` with a dotted callable, following the existing entry conventions.
- [ ] It reads session transcripts within the 30-day retention window and distills them before any model call; a distillation ratio is logged per run.
- [ ] Per-transcript scoring returns strict JSON in which **every** claim carries a verbatim transcript quote; claims without quotes are dropped, not repaired.
- [ ] Scoring results are cached, keyed on transcript content hash + instruction-surface hash, so a re-run costs no model calls when nothing changed.
- [ ] A persistent gap ledger accumulates sightings across runs. A gap is eligible only at **≥2 distinct sessions**; a removal proposal additionally requires harm-class evidence from ≥2 sessions.
- [ ] Rejected proposals are recorded and not re-proposed absent materially new evidence.
- [ ] Proposals are written to a staging copy, never in place, and land as a **PR on a dedicated branch** — never a direct commit to `main`.
- [ ] The PR body shows, per edit: the diff, the supporting evidence quotes, their session sources, and the resulting token-budget delta for the touched file.
- [ ] Only paths under `docs/sdlc/` are ever modified. A guard rejects any proposed edit outside that prefix, and a test asserts it — including for `.claude/skills-global/`.
- [ ] After a proposed edit is applied, `reflections.auditing.run_skills_audit` still passes.
- [ ] Run lock, per-finding dedup, and a one-PR-per-day cap are in place; a paired sweeper closes stale branches/PRs.
- [ ] The write path is off by default (env-gated), and a dry run is demonstrated in the PR description.
- [ ] `reflections/utilities.py::is_high_confidence` is **unchanged**; the non-`code_bug` categories reach the new reflection via a shared evidence stream rather than through the auto-filer.
- [ ] A feature doc under `docs/features/` explains the loop, the evidence thresholds, and the `docs/sdlc`-only boundary, and is linked from `docs/features/README.md`.

## Open questions for `/do-plan`

- Should the evidence stream be a new persisted artifact, or should `session_intelligence` and this reflection both call a shared scorer? The former decouples schedules; the latter avoids double-paying for model calls.
- Is nightly the right cadence, or should it be weekly? A ≥2-session threshold on a nightly run against a quiet day may never accumulate; a weekly run has more evidence per pass and produces fewer PRs to review.
- Should `CLAUDE.md` be in scope as an edit target? It is instruction surface and it is repo-specific, but it is also regex-parsed into worker system prompts (`## Work Completion Criteria`) and asserted byte-for-byte by fixtures. Default recommendation: **out of scope for v1.**

---

This issue will be consumed by `/do-plan` to produce a plan document at `docs/plans/{slug}.md`, then executed by `/do-build`.
